qt/scientific_interfaces/Muon/ALCDataLoadingModel.cpp

```diff
@@ -0,0 +1,346 @@
+                  returnVal.end());
+
+  // Return run number as int (removes leading 0's)
+  return std::stoi(returnVal);
+}
+
+std::string ALCDataLoadingModel::getPathFromFiles(std::vector<std::string> files) {
+  if (files.empty())
+    return "";
+  const auto firstDirectory = files[0u].substr(0u, files[0u].find_last_of("/\\"));
+  // Lambda to compare directories from a path
+  const auto hasSameDirectory = [&firstDirectory](const auto &path) {
+    return path.substr(0u, path.find_last_of("/\\")) == firstDirectory;
+  };
+  const auto sameDirectory = std::all_of(files.cbegin(), files.cend(), hasSameDirectory);
+  return sameDirectory ? firstDirectory : "Multiple Directories";
+}
+
+/**
+ * If custom grouping is supplied, check all detector numbers are valid
+ * @returns :: True if grouping OK, false if bad
+ */
+bool ALCDataLoadingModel::checkCustomGrouping(const std::string &detGroupingType, const std::string &forwardGrouping,
+                                              const std::string &backwardGrouping) {
+  bool groupingOK = true;
+  if (detGroupingType == "Custom") {
+    auto detectors = Mantid::Kernel::Strings::parseRange(isCustomGroupingValid(forwardGrouping, groupingOK));
+    const auto backward = Mantid::Kernel::Strings::parseRange(isCustomGroupingValid(backwardGrouping, groupingOK));
+    if (!groupingOK) {
+      return false;
+    }
+    detectors.insert(detectors.end(), backward.begin(), backward.end());
+    if (std::any_of(detectors.cbegin(), detectors.cend(),
+                    [this](const auto det) { return det < 0 || det > static_cast<int>(m_numDetectors); })) {
+      groupingOK = false;
+    }
+  }
+  return groupingOK;
+}
+/**
+ * Check basic group string is valid
+ * i.e. does not contain letters or start with , or -
+ * @param group :: the string of the grouping
+ * @param isValid :: bool to say if the string is valid
+ * @returns :: True if grouping OK, false if bad
+ */
+std::string ALCDataLoadingModel::isCustomGroupingValid(const std::string &group, bool &isValid) {
+  if (!std::isdigit(static_cast<unsigned char>(group[0])) ||
+      std::any_of(std::begin(group), std::end(group), ::isalpha) ||
+      std::any_of(std::begin(group), std::end(group), is_decimal)) {
+    isValid = false;
+    return "";
+  }
+  return group;
+}
+
+void ALCDataLoadingModel::updateAutoLoadCancelled() {
+  m_lastRunLoadedAuto = -2; // Ensures negative if +1 to not be valid
+  m_wasLastAutoRange = false;
+}
+
+std::string &ALCDataLoadingModel::getRunsText() { return m_runsText; }
+
+void ALCDataLoadingModel::setDirectoryChanged(bool hasDirectoryChanged) { m_directoryChanged = hasDirectoryChanged; }
+
+void ALCDataLoadingModel::setFilesToLoad(const std::vector<std::string> &files) { m_filesToLoad = files; }
+/**
+ * This timer runs every second when we are watching a directory.
+ * If any changes have occurred in the meantime, reload.
+ */
+bool ALCDataLoadingModel::loadFilesFromWatchingDirectory(const std::string &firstFile,
```

Comment:
This function is quite large with a lot of nested if statements. Would it be possible to refactor it to use the "return early" coding style? https://dev.to/jpswade/return-early-12o5#:~:text=The%20motivation%20behind%20this%20is,to%20follow%20the%20code%20low.

I find this improves code readability hugely!